modules/apps/dynamic-data-mapping/dynamic-data-mapping-test/src/testIntegration/java/com/liferay/dynamic/data/mapping/internal/upgrade/v1_0_1/test/DDMFormPortletPreferencesUpgradeProcessTest.java

```diff
@@ -0,0 +1,139 @@
+import com.liferay.portal.kernel.upgrade.UpgradeProcess;
+import com.liferay.portal.kernel.util.PortletKeys;
+import com.liferay.portal.test.log.LogCapture;
+import com.liferay.portal.test.log.LoggerTestUtil;
+import com.liferay.portal.test.rule.Inject;
+import com.liferay.portal.test.rule.LiferayIntegrationTestRule;
+import com.liferay.portal.upgrade.registry.UpgradeStepRegistrator;
+import com.liferay.portal.upgrade.test.util.UpgradeTestUtil;
+
+import java.util.Map;
+
+import javax.portlet.PortletPreferences;
+
+import org.junit.Assert;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/**
+ * @author Paulo Albuquerque
+ */
+@RunWith(Arquillian.class)
+public class DDMFormPortletPreferencesUpgradeProcessTest {
+
+	@ClassRule
+	@Rule
+	public static final LiferayIntegrationTestRule liferayIntegrationTestRule =
+		new LiferayIntegrationTestRule();
+
+	@Test
+	public void testUpgrade() throws Exception {
+		Group group = GroupTestUtil.addGroup();
+
+		DDMFormInstance ddmFormInstance =
+			DDMFormInstanceTestUtil.addDDMFormInstance(
+				GroupTestUtil.addGroup(), TestPropsValues.getUserId());
+
+		Layout layout = LayoutTestUtil.addTypeContentLayout(group);
+
+		String portletId = LayoutTestUtil.addPortletToLayout(
+			layout, DDMPortletKeys.DYNAMIC_DATA_MAPPING_FORM);
+
+		PortletPreferences portletPreferences =
+			LayoutTestUtil.getPortletPreferences(layout, portletId);
+
+		portletPreferences.setValue(
+			"formInstanceId",
+			String.valueOf(ddmFormInstance.getFormInstanceId()));
+		portletPreferences.setValue(
+			"groupId", String.valueOf(group.getGroupId()));
+
+		_portletPreferencesLocalService.updatePreferences(
+			PortletKeys.PREFS_OWNER_ID_DEFAULT,
+			PortletKeys.PREFS_OWNER_TYPE_LAYOUT, layout.getPlid(), portletId,
+			portletPreferences);
+
+		portletPreferences = LayoutTestUtil.getPortletPreferences(
+			layout, portletId);
+
+		Map<String, String[]> preferencesMap = portletPreferences.getMap();
+
+		String[] preference = preferencesMap.get("formInstanceId");
+
+		Assert.assertEquals(
+			String.valueOf(ddmFormInstance.getFormInstanceId()), preference[0]);
+
+		preference = preferencesMap.get("groupId");
+
+		Assert.assertEquals(String.valueOf(group.getGroupId()), preference[0]);
+
+		try (LogCapture logCapture = LoggerTestUtil.configureLog4JLogger(
+				_CLASS_NAME, LoggerTestUtil.OFF)) {
+
+			UpgradeProcess upgradeProcess = UpgradeTestUtil.getUpgradeStep(
+				_upgradeStepRegistrator, _CLASS_NAME);
+
+			upgradeProcess.upgrade();
+
+			_multiVMPool.clear();
+		}
+
+		portletPreferences = LayoutTestUtil.getPortletPreferences(
+			layout, portletId);
+
+		preferencesMap = portletPreferences.getMap();
+
+		DDMStructure ddmStructure = _ddmStructureLocalService.getDDMStructure(
+			ddmFormInstance.getStructureId());
+
+		preference = preferencesMap.get("ddmStructureExternalReferenceCode");
+
+		Assert.assertEquals(
```

Comment:
@p-albuquerque Here we have something like this:

```
portletPreferences = ...

preferencesMap =  uses the portletPreferences

DDMStructure ddmStructure = ...

preference = uses the preferencesMap

```

Do you see how the ddmStructure is suddenly between two related variables?

In this case I would try to escape creating all these variables, either by using inline methods like `MapUtil.get` or by creating private methods, as I suggested above.